src/bar.cpp

```diff
@@ -683,6 +710,59 @@ void waybar::Bar::setVisible(bool value) {
 
 void waybar::Bar::toggle() { setVisible(!visible); }
 
+auto waybar::Bar::showMainbar(GtkWidget* widget, const gchar* h, gpointer user_data) -> void {
+  auto curbar = static_cast<Bar*>(user_data);
+  curbar->autohide_mutex.lock();
+
+  // Delete timeout callbacks;
+  auto tag = curbar->autohide_timeout_handle.load();
+  if(tag !=0 && g_source_remove(tag)) {
+    spdlog::debug("removed autohide timeout handle in showMainbar: {}",tag);
+    curbar->autohide_timeout_handle.store(0);
+  }
+
+  spdlog::debug("hotspot: off; bar: on");
+  if(!curbar->window.is_visible()) curbar->window.show_all();
+  if(curbar->hotspotWindow.is_visible()) curbar->hotspotWindow.hide();
+
+  curbar->autohide_mutex.unlock();
+}
+
+auto waybar::Bar::hideMainbarCallback(gpointer user_data) -> void {
+  auto curbar = static_cast<Bar*>(user_data);
+  curbar->autohide_mutex.lock();
+
+  spdlog::debug("hotspot: on; bar: off");
+  if(!curbar->hotspotWindow.is_visible()) curbar->hotspotWindow.show_all();
+  if(curbar->window.is_visible())curbar->window.hide();
+
+  curbar->autohide_timeout_handle.store(0);
+  curbar->autohide_mutex.unlock();
+}
+
+auto waybar::Bar::hideMainbar(GtkWidget* widget, const char* h, gpointer user_data) -> void {
+  auto curbar = static_cast<Bar*>(user_data);
+  curbar->autohide_mutex.lock();
+  auto tag = curbar->autohide_timeout_handle.load();
+  if(tag !=0 && g_source_remove(tag)) {
+    spdlog::debug("removed timeout handle in hideMainbar: {}",tag);
+    curbar->autohide_timeout_handle.store(0);
+  } else {
+    curbar->autohide_timeout_handle.store(g_timeout_add_once(curbar->autohide_delay_ms.load(),GSourceOnceFunc(hideMainbarCallback),curbar));
+  }
+  curbar->autohide_mutex.unlock();
+}
+
+void waybar::Bar::setupAutohide() {
+  this->autohide_timeout_handle.store(0);
+  this->autohide_delay_ms.store(config["autohide-delay"].isUInt() ? config["autohide-delay"].asUInt() : 0);
+
+  gtk_layer_set_layer(hotspotWindow.gobj(), GTK_LAYER_SHELL_LAYER_OVERLAY);
```

Comment:
gtk-layer-shell is an optional dependency, and Waybar can be compiled without it. Even if it doesn't make much sense nowadays.

Reply:
So, should i enable autohide when only waybar is compiled with gtk-layer-shell?

Reply:
@alebastr , done.